draco/asp/soft.lp

```diff
@@ -30,18 +30,67 @@ preference(same_field,F) :-
     entity(mark,_,M),
     { entity(encoding,M,E): attribute((encoding,field),E,F) } = 2.
 
-% @soft(type_linear) linear scale
-preference(type_linear,E) :-
+% @soft(same_field_grt3) Prefer not to use the same field three or more times for the same mark.
+preference(same_field_grt3,F) :-
+    entity(field,_,F),
+    entity(mark,_,M),
+    { entity(encoding,M,E): attribute((encoding,field),E,F) } >= 3.
+
+% @soft(count_grt1) Prefer not to use count more than once.
+preference(count_grt1,M) :-
+    entity(mark,_,M),
+    { entity(encoding,M,E): attribute((encoding,aggregate),E,count) } > 1.
+
+% @soft(shape_high_cardinality) Shape channel should not have too high cardinality.
+preference(shape_high_cardinality,E) :-
+    attribute((encoding,channel),E,shape),
+    attribute(encoding_cardinality,E,N),
+    N > 5.
+
+% @soft(number_categorical) Numbers should not be nominal.
+preference(number_categorical,F) :-
+    attribute((field,type),F,number),
+    attribute(field_scale,F,categorical).
+
+% @soft(bin_number_low_cardinality) Binned number field should not have too low cardinality.
+preference(bin_number_low_cardinality,E) :-
+    attribute((field,type),F,number),
+    attribute((encoding,field),E,F),
+    attribute((encoding,binning),E,B),
+    B < 15.
```

Comment:
also I think this constraint should be about data cardinality(unique) rather than the number of bins